### ISSUE TITLE
op_ref false positive

### ISSUE THREAD

**Issue by tspiteri** [Manishearth/rust-clippy#1704]
In the code below, the operator is defined for both moving and borrowing the rhs. The lhs is moved in both implementations. The code triggers a false positive of the `op_ref` lint.

```
use std::ops::BitAnd;
struct X(i32);
impl BitAnd for X {
    type Output = X;
    fn bitand(self, rhs: X) -> X {
        X(self.0 & rhs.0)
    }
}
impl<'a> BitAnd<&'a X> for X {
    type Output = X;
    fn bitand(self, rhs: &'a X) -> X {
        X(self.0 & rhs.0)
    }
}
pub fn foo() -> i32 {
    let x = X(1);
    let y = X(2);
    let z = x & &y;
    z.0
}
```

**Comment by oli-obk** [comment]
hmm now that I think about it. We should only be emitting this lint for Copy types.

**Comment by tspiteri** [comment]
In this case, if `X` is `Copy` there is still a false positive, as `x` in `x & &y` is not a reference and cannot be dereferenced as the lint suggests. Though then the second implementation would be unnecessary.

**Comment by oli-obk** [comment]
Yea, the deref suggestion shouldn't show up in the Copy case.
